### PR TITLE
Texture2D Array and Texture 3D

### PR DESCRIPTION
**Pull request by Kink3d**

### CODE REVIEW

File: ShaderGraph/com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/SampleTexture2DArrayNode.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                , edgesSampler.Any() ? GetSlotValue(SamplerInput, generationMode) : "sampler" + id
+                , uvName
+                , indexName);
+
+            visitor.AddShaderChunk(result, true);
+
+            visitor.AddShaderChunk(string.Format("{0} {1} = {2}.r;", precision, GetVariableNameForSlot(OutputSlotRId), GetVariableNameForSlot(OutputSlotRGBAId)), true);
+            visitor.AddShaderChunk(string.Format("{0} {1} = {2}.g;", precision, GetVariableNameForSlot(OutputSlotGId), GetVariableNameForSlot(OutputSlotRGBAId)), true);
+            visitor.AddShaderChunk(string.Format("{0} {1} = {2}.b;", precision, GetVariableNameForSlot(OutputSlotBId), GetVariableNameForSlot(OutputSlotRGBAId)), true);
+            visitor.AddShaderChunk(string.Format("{0} {1} = {2}.a;", precision, GetVariableNameForSlot(OutputSlotAId), GetVariableNameForSlot(OutputSlotRGBAId)), true);
```
Comment (pbbastian) [inline ShaderGraph/com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/SampleTexture2DArrayNode.cs]: None of these should be unique

File: ShaderGraph/com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/SampleTexture3DNode.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+            var id = GetSlotValue(TextureInputId, generationMode);
+            var result = string.Format("{0}4 {1} = SAMPLE_TEXTURE3D({2}, {3}, {4});"
+                , precision
+                , GetVariableNameForSlot(OutputSlotId)
+                , id
+                , edgesSampler.Any() ? GetSlotValue(SamplerInput, generationMode) : "sampler" + id
+                , uvName);
+
+            visitor.AddShaderChunk(result, true);
```
Comment (pbbastian) [inline ShaderGraph/com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/SampleTexture3DNode.cs]: shouldn't be using unique

File: ShaderGraph/com.unity.shadergraph/Editor/Drawing/Manipulators/GraphDropTarget.cs [hunk 3]
```diff
@@ -65,6 +65,38 @@ void CreateNode(object obj, Vector2 nodePosition)
                     inputslot.texture = texture2D;
             }
 
+            var textureArray = obj as Texture2DArray;
```
Comment (pbbastian) [inline ShaderGraph/com.unity.shadergraph/Editor/Drawing/Manipulators/GraphDropTarget.cs]: Needs to be rewritten into the new drag and drop code

File: ShaderGraph/com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/SampleTexture2DArrayNode.cs (hunk 1 above)
Comment (pbbastian) [reply ShaderGraph/com.unity.shadergraph/Editor/Data/Nodes/Input/Texture/SampleTexture2DArrayNode.cs]: Seems to be an issue that was copied across. Will fix in another PR.

**Comment by JapaMala** [comment]
Just out of curiosity, was this done incorporating https://github.com/Unity-Technologies/ShaderGraph/pull/388 or independently?

**Pull request merged**